Propagate extras to media item

snipd-mikel opened the pull request:
- **Updated androidx.media** to support new MediaItem metadata like [DESCRIPTION_EXTRAS_KEY_COMPLETION_PERCENTAGE](https://developer.android.com/reference/androidx/media/utils/MediaConstants#DESCRIPTION_EXTRAS_KEY_COMPLETION_PERCENTAGE)
- **Propagate dart's mediaItem's extras to `MediaItemCompat` description's extras.** Until now the `MediaItemCompat` was created using the `MediaMedataCompat` description. However, when building a `MediaMetadataCompat` instance, it doesn't pass all the extras to its `MediaDescriptionCompat` object.
- **Update the progress bar in the browse view as content is playing**: Set the [PLAYBACK_STATE_EXTRAS_KEY_MEDIA_ID](https://developer.android.com/reference/androidx/media/utils/MediaConstants#PLAYBACK_STATE_EXTRAS_KEY_MEDIA_ID) on the playback state so the MediaService knows about which item is currently playing and updates the Browse view accordingly.

#941 

## Pre-launch Checklist

- [X] I read the [CONTRIBUTING.md] and followed the process outlined there for submitting PRs.
- [X] My change is not breaking and lands in `minor` branch OR my change is breaking and lands in `major` branch.
- [X] If I'm the first to contribute to the next version, I incremented the version number in `pubspec.yaml` according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change (format: `* DESCRIPTION OF YOUR CHANGE (@your-git-username)`).
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I ran `dart analyze`.
- [X] I ran `dart format`.
- [X] I ran `flutter test` and all tests are passing.

<!-- Please consider also adding unit tests covering your new code. -->

<!-- Links -->
[CONTRIBUTING.md]: https://github.com/ryanheise/audio_service/blob/minor/CONTRIBUTING.md#making-a-pull-request
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

snipd-mikel committed:
Update androidx.meida version

snipd-mikel committed:
Propagate dart extras to MediaItem description extras

snipd-mikel committed:
Link current playing item to current playback state

snipd-mikel committed:
Update CHANGELOG and bump version to 0.18.6

ryanheise commented on audio_service/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+* Android: Update androidx.media to 1.6.0 to support extra MediaItem metadata options (@snipd-mikel)
+* Android: Propagate mediaItem extras coming from dart to MediaItem's MediaDescription extras (@snipd-mikel)
+* Android: Update the progress bar in the browse view as content is playing as explained [here](https://developer.android.com/training/cars/media#browse-progress-bar) (@snipd-mikel)
+
```
Can we make these entries more brief to fit in a line, like the others?

* Migrate to androidx.media 1.6.0 (@snipd-mikel)
* Propagate MediaItem extras to Android Auto (@snipd-mikel)
* Update progress bar in Android Auto (@snipd-mikel)

ryanheise commented on audio_service/android/src/main/java/com/ryanheise/audioservice/AudioServicePlugin.java:
```diff
@@ -1115,12 +1112,61 @@ private static MediaMetadataCompat createMediaMetadata(Map<?, ?> rawMediaItem) {
+            }
+        }
+        return new MediaDescriptionCompat.Builder()
+                .setTitle(description.getTitle())
```
Missing setSubtitle()

ryanheise commented on audio_service/android/src/main/java/com/ryanheise/audioservice/AudioServicePlugin.java:
```diff
@@ -1115,12 +1112,61 @@ private static MediaMetadataCompat createMediaMetadata(Map<?, ?> rawMediaItem) {
+        for (Object o : extras.keySet()) {
+            String key = (String)o;
+            final Object value = extras.get(key);
+            if (value instanceof Long) {
```
Use `AudioServicePlugin.mapToBundle()`.

ryanheise commented:
Hi @snipd-mikel , I have added some comments within the code.

Would you also be able to modify `example/lib/example_multiple_handlers.dart` to demonstrate this? For example, do you need to use `DESCRIPTION_EXTRAS_KEY_COMPLETION_PERCENTAGE`? Or would it be possible for the plugin to automatically manage this?

snipd-mikel commented:
@ryanheise  Thanks for the comments, will work on those changes.

Regarding your point on wether we need to manually set `DESCRIPTION_EXTRAS_KEY_COMPLETION_PERCENTAGE`, my preferred solution would be to introduce a `progress` field on the `MediaItem` from dart. The same could apply to the other extras like `downloadStatus`, `playStatus` or `isExplicit`.

I didn't add that because this will only apply for Android and I wanted to keep the changes to a minimum. But if you think it makes sense to go on that direction, I will be more than happy to include it in the PR. Let me know what you think :)

---- 
On a side note regarding this only applying to Android: I have been working on an Apple Carplay integration for our podcast app. Without going into much detail, the Carplay integration is based on building the UI using templates so it doesn't reuse what is pushed through audio_service.

I have done this using the `flutter_carplay` plugin, where you build the templates on dart. I think it would be possible to reuse `audio_service` data (like queue, media item children, progress or download status) even if the templates have to be built manually. That way, part of the code can be shared between platform. This may also allow in the future to create a new plugin to easily connect `audio_service` to `carplay` as we would have most of the data available there already.

ryanheise commented:
> Regarding your point on wether we need to manually set `DESCRIPTION_EXTRAS_KEY_COMPLETION_PERCENTAGE`, my preferred solution would be to introduce a `progress` field on the `MediaItem` from dart. The same could apply to the other extras like `downloadStatus`, `playStatus` or `isExplicit`.
> 
> I didn't add that because this will only apply for Android and I wanted to keep the changes to a minimum. But if you think it makes sense to go on that direction, I will be more than happy to include it in the PR. Let me know what you think :)

If the plugin can manage it, what I had in mind was to use the existing state parameters. We already have `updatePosition` and `updateTime`. We don't need to add more fields, that's enough to indicate the progress.

As for the others (e.g. `isExplicit`) I am fine to leave them as extras because what I'm talking about here with progress is to leverage the existing fields to achieve this functionality, not adding a new field.

> On a side note regarding this only applying to Android: I have been working on an Apple Carplay integration for our podcast app. Without going into much detail, the Carplay integration is based on building the UI using templates so it doesn't reuse what is pushed through audio_service.
> 
> I have done this using the `flutter_carplay` plugin, where you build the templates on dart. I think it would be possible to reuse `audio_service` data (like queue, media item children, progress or download status) even if the templates have to be built manually. That way, part of the code can be shared between platform. This may also allow in the future to create a new plugin to easily connect `audio_service` to `carplay` as we would have most of the data available there already.

I'd be happy to have Carplay support directly in audio_service, although it will most likely need a contributor to kick things off, since the Car side of things is not something I directly use myself. I would suggest creating a new feature request describing what you have in mind so that people can start thinking about how to contribute.

snipd-mikel commented:
> If the plugin can manage it, what I had in mind was to use the existing state parameters. We already have updatePosition and updateTime. We don't need to add more fields, that's enough to indicate the progress.

The `DESCRIPTION_EXTRAS_KEY_COMPLETION_PERCENTAGE` is actually applied per MediaItem, so it shows it's progress even if it is not playing (it can be used to show where the user left playback of an episode). Because of that we cannnot use `updatePosition` as it is part of the PlaybackState.

I can add `DESCRIPTION_EXTRAS_KEY_COMPLETION_PERCENTAGE` to the example project or other extras like the download status to demonstrate its usage.

> I'd be happy to have Carplay support directly in audio_service, although it will most likely need a contributor to kick things off, since the Car side of things is not something I directly use myself. I would suggest creating a new feature request describing what you have in mind so that people can start thinking about how to contribute.

I'll be happy to create that feature request and what my experience has been integrating with AppleCarplay with the other plugin.

I have already addressed your comments on the PR.

ryanheise commented:
I see, that makes sense. I guess I could imagine the plugin automatically setting the completion percentage based on the last known position for each media item, but that would probably be doing too much, making assumptions for how the app will actually work. E.g. whether or not jumping back to a previous track will resume from its last known position, or whether it will start over from the beginning. So in that case, it makes sense as you suggested to demonstrate completion_percentage in the example.

ryanheise commented:
My apologies, I just introduced a merge conflict in CHANGELOG.md (It was my fault for not getting the last entry correct).

ryanheise commented:
I can try to merge and resolve the conflicts without an example, and if needed an example and documentation can be added later.

ryanheise commented:
All merged :+1: Thanks for your work on this contribution.

snipd-mikel commented:
Hi @ryanheise thanks a lot for finishing up the PR changes, got busy last weeks and couldn't find time to it
